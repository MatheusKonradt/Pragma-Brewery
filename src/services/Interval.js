class Interval {
  /**
   * @param {number} a
   * @param {number} b
   */
  constructor(a, b) {
    this.a = a;
    this.b = b;
    if (a > b) throw new Error('invalid interval.');
  }

  /**
   * @return {number}
   */
  getSize() {
    return this.b - this.a;
  }

  /**
   * @param {number} number
   * @param {Interval} interval
   * @return {number}
   */
  scaleNumberToOtherInterval(number, interval) {
    const ratio = interval.getSize() / this.getSize();
    return (ratio * number) - (ratio * this.a) + interval.a;
  }

  /**
   * @param {number} step
   * @param {Function} cb
   * @return {Array<any>}
   */
  map(step, cb) {
    const values = [];
    for(let i=this.a ; i<this.b ; i+= step) values.push(i);
    return values.map(cb);
  }
}

export default Interval;