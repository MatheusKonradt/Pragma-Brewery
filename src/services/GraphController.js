import _ from 'lodash';
import Coord from './Coord';
import Interval from './Interval';
import CoordSpace from './CoordSpace';

class GraphController {
  /**
   * @param {HTMLCanvasElement} $canvas
   */
  constructor($canvas) {
    if ($canvas) this.setCanvas($canvas);
    this.values = [];
    this.setMaxDisplayedValues(15);
    this.setMarginOffset(0.3);
  }

  /**
   * @param {HTMLCanvasElement} $canvas
   * @return {GraphController}
   */
  setCanvas($canvas) {
    this.$canvas = $canvas;
    this.ctx = $canvas.getContext('2d');
    this.refreshCanvasSpace();
    return this;
  }

  /**
   * @param {number} val
   * @return {GraphController}
   */
  setMarginOffset(val) {
    this.marginOffset = val;
    return this;
  }

  /**
   * @param {Interval} val
   * @return {GraphController}
   */
  setXInterval(val) {
    this.xInterval = val;
    this.refreshVirtualSpace();
    return this;
  }

  /**
   * @param {Interval} val
   * @return {GraphController}
   */
  setYInterval(val) {
    this.yInteval = val;
    this.refreshVirtualSpace();
    return this;
  }

  /**
   * @param {Interval} val
   * @return {GraphController}
   */
  setMinYInterval(val) {
    this.minYInterval = val;
  }

  /**
   * @param val
   * @return {GraphController}
   */
  setMaxDisplayedValues(val) {
    this.maxDisplayedValues = val;
    return this;
  }

  /**
   * @return {GraphController}
   */
  refreshCanvasSpace() {
    this.xCanvasInterval = new Interval(0, this.$canvas.width);
    this.yCanvasInterval = new Interval(0, this.$canvas.height);
    this.canvasCoordSpace = new CoordSpace(this.xCanvasInterval, this.yCanvasInterval);
    return this;
  }

  /**
   * @param {number} val
   * @return {GraphController}
   */
  addValue(val) {
    this.values.push(val);
    if (this.values.length > this.maxDisplayedValues) {
      this.values.shift();
    }
    return this;
  }

  /**
   * @return {GraphController}
   */
  refreshVirtualSpace() {
    this.virtualCoordSpace = new CoordSpace(this.xInterval, this.yInteval);
    return this;
  }

  /**
   * @return {GraphController}
   */
  updateYIntervalUsingHistoricValues() {
    const maxHistoricValue = _.max(this.values);
    const minHistoricValue = _.min(this.values);
    const a = Math.min(minHistoricValue, this.minYInterval.a) - this.marginOffset;
    const b = Math.max(maxHistoricValue, this.minYInterval.b) + this.marginOffset;
    this.setYInterval(new Interval(a, b));
    return this;
  }

  /**
   * @param {Coord} from
   * @param {Coord} to
   * @param {number} lineWidth
   * @param {string} strokeStyle
   * @return {GraphController}
   */
  drawLine(from, to, lineWidth, strokeStyle) {
    this.ctx.beginPath();
    this.ctx.lineWidth = lineWidth;
    this.ctx.strokeStyle = strokeStyle;
    this.ctx.moveTo(from.x, from.y);
    this.ctx.lineTo(to.x, to.y);
    this.ctx.stroke();
    return this;
  }

  /**
   * @return {GraphController}
   */
  clearCanvas() {
    this.ctx.clearRect(
      this.xCanvasInterval.a, this.yCanvasInterval.a,
      this.xCanvasInterval.b, this.yCanvasInterval.b,
    );
    return this;
  }

  /**
   * @return {GraphController}
   */
  drawFixedLines() {
    const step = this.yInteval.getSize() / 10;
    this.yInteval.map(step, (y) => {
      const fromCoord = new Coord(this.xInterval.a, y);
      fromCoord.setSpace(this.virtualCoordSpace);
      fromCoord.transformSpace(this.canvasCoordSpace);

      const toCoord = new Coord(this.xInterval.b, y);
      toCoord.setSpace(this.virtualCoordSpace);
      toCoord.transformSpace(this.canvasCoordSpace);

      this.drawLine(fromCoord, toCoord, 1, '#fff');
    });
    return this;
  }

  /**
   * @return {Coord[]}
   */
  getValuesCoords() {
    return this.values.map((value, i) => {
      const coord = new Coord(i, value);
      coord.setSpace(this.virtualCoordSpace);
      coord.transformSpace(this.canvasCoordSpace);
      coord.invertYAxis();
      return coord;
    });
  }

  /**
   * @return {GraphController}
   */
  drawLimitLines() {
    let fromCoord = new Coord(this.xInterval.a, this.minYInterval.a);
    fromCoord
      .setSpace(this.virtualCoordSpace)
      .transformSpace(this.canvasCoordSpace)
      .invertYAxis();

    let toCoord = new Coord(this.xInterval.b, this.minYInterval.a);
    toCoord
      .setSpace(this.virtualCoordSpace)
      .transformSpace(this.canvasCoordSpace)
      .invertYAxis();

    this.drawLine(fromCoord, toCoord, 2, '#f00');

    fromCoord = new Coord(this.xInterval.a, this.minYInterval.b);
    fromCoord
      .setSpace(this.virtualCoordSpace)
      .transformSpace(this.canvasCoordSpace)
      .invertYAxis();

    toCoord = new Coord(this.xInterval.b, this.minYInterval.b);
    toCoord
      .setSpace(this.virtualCoordSpace)
      .transformSpace(this.canvasCoordSpace)
      .invertYAxis();

    this.drawLine(fromCoord, toCoord, 2, '#f00');

    return this;
  }

  /**
   * @return {GraphController}
   */
  drawValuesLine() {
    const coords = this.getValuesCoords();
    let currentCoord = _.first(coords);
    coords.forEach((coord) => {
      this.drawLine(currentCoord, coord, 2, '#000');
      currentCoord = coord;
    });
    return this;
  }

  /**
   * @return {GraphController}
   */
  draw() {
    this.updateYIntervalUsingHistoricValues();
    this.clearCanvas();
    this.drawFixedLines();
    this.drawLimitLines();
    this.drawValuesLine();
    return this;
  }
}

export default GraphController;
