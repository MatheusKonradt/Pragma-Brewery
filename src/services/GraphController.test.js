import GraphController from './GraphController';
import Interval from './Interval';

test('should map values to coordinates', () => {
  const controller = new GraphController(null);
  controller.setYInterval(new Interval(3, 8));
  controller.setXInterval(new Interval(0, 10));
  controller.$canvas = { width: 300, height: 150 };
  controller.refreshCanvasSpace();
  controller
    .addValue(4)
    .addValue(5)
    .addValue(3)
    .addValue(9);
  const coords = controller.getValuesCoords();
  expect(coords).toMatchSnapshot();
});

test('should recreate canvas space', () => {
  const controller = new GraphController(null);
  controller.$canvas = { width: 300, height: 150 };
  controller.refreshCanvasSpace();
  expect(controller.canvasCoordSpace).toMatchSnapshot();
});

test('should respect maxDisplayedValues when adding new values', () => {
  const controller = new GraphController(null);
  controller.setMaxDisplayedValues(4);
  controller
    .addValue(4)
    .addValue(5)
    .addValue(3)
    .addValue(8)
    .addValue(9);
  expect(controller.values.length).toBe(4);
  expect(controller.values).toMatchSnapshot();
});

test('should update graph scale acordingly to max and min value', () => {
  const controller = new GraphController(null);
  controller.values = [1, 3, 4, 6, 7, -1, 3, 8]; // MAX is 8, MIN is -1
  controller.setMinYInterval(new Interval(3, 5)); // Should not affect result since this interval is contained in [-1, 8]
  controller.updateYIntervalUsingHistoricValues();
  expect(controller.yInteval).toMatchSnapshot();
});
