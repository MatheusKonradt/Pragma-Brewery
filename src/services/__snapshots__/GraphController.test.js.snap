// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should map values to coordinates 1`] = `
Array [
  Coord {
    "space": CoordSpace {
      "xInterval": Interval {
        "a": 0,
        "b": 300,
      },
      "yInterval": Interval {
        "a": 0,
        "b": 150,
      },
    },
    "x": 0,
    "y": 120,
  },
  Coord {
    "space": CoordSpace {
      "xInterval": Interval {
        "a": 0,
        "b": 300,
      },
      "yInterval": Interval {
        "a": 0,
        "b": 150,
      },
    },
    "x": 30,
    "y": 90,
  },
  Coord {
    "space": CoordSpace {
      "xInterval": Interval {
        "a": 0,
        "b": 300,
      },
      "yInterval": Interval {
        "a": 0,
        "b": 150,
      },
    },
    "x": 60,
    "y": 150,
  },
  Coord {
    "space": CoordSpace {
      "xInterval": Interval {
        "a": 0,
        "b": 300,
      },
      "yInterval": Interval {
        "a": 0,
        "b": 150,
      },
    },
    "x": 90,
    "y": -30,
  },
]
`;

exports[`should recreate canvas space 1`] = `
CoordSpace {
  "xInterval": Interval {
    "a": 0,
    "b": 300,
  },
  "yInterval": Interval {
    "a": 0,
    "b": 150,
  },
}
`;

exports[`should respect maxDisplayedValues when adding new values 1`] = `
Array [
  5,
  3,
  8,
  9,
]
`;

exports[`should update graph scale acordingly to max and min value 1`] = `
Interval {
  "a": -1.3,
  "b": 8.3,
}
`;
